opa/pkg/main.go

```diff
@@ -185,6 +192,133 @@ func processCheckTarget(ctx context.Context, check customCheck, checkInstance cu
 	return actionItems, nil
 }
 
+func getInsightsChecks() (clusterCheckModel, error) {
+	var jsonResponse clusterCheckModel
+
+	url := os.Getenv("FAIRWINDS_INSIGHTS_HOST") + "/v0/organizations/" + os.Getenv("FAIRWINDS_ORG") + "/clusters/" + os.Getenv("FAIRWINDS_CLUSTER") +
+		"/data/opa/customChecks"
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return jsonResponse, err
+	}
+	req.Header.Set("Authorization", "Bearer "+os.Getenv("FAIRWINDS_TOKEN"))
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return jsonResponse, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return jsonResponse, fmt.Errorf("failed to retrieve updated checks with a status code of : %d", resp.StatusCode)
+	}
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return jsonResponse, err
+	}
+	err = json.Unmarshal(responseBody, &jsonResponse)
+	if err != nil {
+		return jsonResponse, err
+	}
+	return jsonResponse, nil
+}
+
+func refreshLocalChecks(ctx context.Context, dynamicInterface dynamic.Interface) error {
+
+	thisNamespace := "insights-agent"
+	namespaceBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err == nil { //Ignore errors because that means this isn't running in a container
+		thisNamespace = string(namespaceBytes)
+	}
+	checkInstances, err := dynamicInterface.Resource(instanceGvr).Namespace(thisNamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	checks, err := dynamicInterface.Resource(checkGvr).Namespace(thisNamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	jsonResponse, err := getInsightsChecks()
+	if err != nil {
+		return err
+	}
+
+	for _, check := range checks.Items {
+		found := false
+		for _, supposedCheck := range jsonResponse.Checks {
+			if check.GetName() == supposedCheck.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			err = dynamicInterface.Resource(checkGvr).Namespace(check.GetNamespace()).Delete(ctx, check.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+
+		}
+	}
+	for _, instance := range checkInstances.Items {
+		found := false
+		for _, supposedInstance := range jsonResponse.Instances {
+			if instance.GetName() == supposedInstance.AdditionalData.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			err = dynamicInterface.Resource(instanceGvr).Namespace(instance.GetNamespace()).Delete(ctx, instance.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for _, supposedCheck := range jsonResponse.Checks {
+		found := false
+		for _, check := range checks.Items {
+			if check.GetName() == supposedCheck.Name {
+				found = true
+				break
+			}
+		}
+		newCheck := supposedCheck.GetUnstructuredObject(thisNamespace)
+		if found {
+			err = dynamicInterface.Resource(checkGvr).Namespace(thisNamespace).Delete(ctx, supposedCheck.Name, metav1.DeleteOptions{})
```

Comment:
couldn't we start by just deleting _all_ the checks/instances, and then replacing them with what we saw in Insights?

Reply:
I didn't want to do that because if something goes wrong at some point we can still have most of the checks and continue on. So as we're processing the new checks it deletes the old version of that check.